core/src/test/java/org/infinispan/persistence/PersistenceManagerTest.java

```diff
@@ -0,0 +1,88 @@
+import org.infinispan.marshall.core.MarshalledEntryImpl;
+import org.infinispan.persistence.dummy.DummyInMemoryStoreConfigurationBuilder;
+import org.infinispan.persistence.manager.PersistenceManager;
+import org.infinispan.test.Exceptions;
+import org.infinispan.test.SingleCacheManagerTest;
+import org.infinispan.test.fwk.CleanupAfterMethod;
+import org.infinispan.test.fwk.TestCacheManagerFactory;
+import org.infinispan.util.concurrent.WithinThreadExecutor;
+import org.testng.annotations.Test;
+
+/**
+ * A {@link PersistenceManager} unit test.
+ *
+ * @author Pedro Ruivo
+ * @since 9.3
+ */
+@Test(groups = "unit", testName = "persistence.PersistenceManagerTest")
+@CleanupAfterMethod
+public class PersistenceManagerTest extends SingleCacheManagerTest {
+
+   public void testProcessAfterStop() {
+      PersistenceManager persistenceManager = extractComponent(cache, PersistenceManager.class);
+      StreamingMarshaller marshaller = extractComponent(cache, StreamingMarshaller.class);
+      persistenceManager.writeToAllNonTxStores(marshalledEntry("k", "v", marshaller), BOTH);
+      //simulates the scenario where, concurrently, the cache is stopping and handling a topology update.
+      persistenceManager.stop();
+      //the org.infinispan.persistence.dummy.DummyInMemoryStore throws an exception if the process() method is invoked after stopped.
+      persistenceManager.processOnAllStores(new WithinThreadExecutor(), null,
+            (marshalledEntry, taskContext) -> fail("shouldn't run"),
+            true, true);
+   }
+
+   public void testStopDuringProcess() throws ExecutionException, InterruptedException, TimeoutException {
+      PersistenceManager persistenceManager = extractComponent(cache, PersistenceManager.class);
+      StreamingMarshaller marshaller = extractComponent(cache, StreamingMarshaller.class);
+      //simulates the scenario where, concurrently, the cache is stopped during a process loop
+      persistenceManager.writeToAllNonTxStores(marshalledEntry("k1", "v1", marshaller), BOTH);
+      persistenceManager.writeToAllNonTxStores(marshalledEntry("k2", "v2", marshaller), BOTH);
+      persistenceManager.writeToAllNonTxStores(marshalledEntry("k3", "v3", marshaller), BOTH);
+      final CountDownLatch before = new CountDownLatch(1);
+      final CountDownLatch after = new CountDownLatch(1);
+      final AtomicInteger count = new AtomicInteger(0);
+      Future<Void> c = fork(() -> persistenceManager.processOnAllStores(new WithinThreadExecutor(), null,
+            (marshalledEntry, taskContext) -> {
+               before.countDown();
+               after.await();
+               count.incrementAndGet();
+            },
+            true, true));
+      before.await(30, TimeUnit.SECONDS);
+      Future<Void> stopFuture = fork(persistenceManager::stop);
+      //stop is unable to proceed while the process isn't finish.
+      Exceptions.expectException(TimeoutException.class, () -> stopFuture.get(15, TimeUnit.SECONDS));
```

Comment:
15 seconds seems awfully long.

Reply:
10 seconds is fine?

Reply:
I would say maybe a few hundred milliseconds at most? This means this test will always take 15 seconds to complete currently.